worker/introspection/leases.go

```diff
@@ -0,0 +1,185 @@
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/juju/juju/core/lease"
+	"github.com/juju/juju/core/raftlease"
+	"gopkg.in/yaml.v2"
+)
+
+type leaseHandler struct {
+	leases Leases
+}
+
+// ServeHTTP is part of the http.Handler interface.
+func (h leaseHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	ss, err := h.leases.Snapshot()
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprintf(w, "error: %v\n", err)
+		return
+	}
+	snapshot, ok := ss.(*raftlease.Snapshot)
+	if !ok {
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprintf(w, "expected *raftlease.Snapshot\n")
+		return
+	}
+	if snapshot.Version != 1 {
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprintf(w, "only understand how to show version 1 snapshots\n")
+		return
+	}
+
+	data := h.translateSnapshot(snapshot)
+
+	q := r.URL.Query()
+	if v := q.Get("model"); v != "" {
+		data = h.filterModel(data, v)
+	}
+	if v := q["app"]; len(v) > 0 {
+		data = h.filterApp(data, v)
+	}
+
+	bytes, err := yaml.Marshal(h.format(data))
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprintf(w, "error: %v\n", err)
+		return
+	}
+	w.Write(bytes)
+}
+
+type leases struct {
+	// TODO: handle pinning
+	controller map[string]leaseInfo
+	models     map[string]map[string]leaseInfo
+}
+
+type leaseInfo struct {
+	holder   string
+	acquired time.Duration
+	expires  time.Duration
+}
+
+func (h leaseHandler) translateSnapshot(snapshot *raftlease.Snapshot) *leases {
+	result := &leases{
+		controller: make(map[string]leaseInfo),
+		models:     make(map[string]map[string]leaseInfo),
+	}
+	now := snapshot.GlobalTime
+	makeLeaseInfo := func(entry raftlease.SnapshotEntry) leaseInfo {
+		acquired := entry.Start.Sub(now).Round(time.Second)
+		return leaseInfo{
+			holder:   entry.Holder,
+			acquired: acquired,
+			// We don't need to round the expiry at this stage because
+			// the entry.Duration is always one minute. Even if this changes
+			// it is unlikely to be a fraction of a second, so we should
+			// still be fine.
+			expires: acquired + entry.Duration,
+		}
+	}
+	for key, value := range snapshot.Entries {
+		switch key.Namespace {
+		case lease.SingularControllerNamespace:
+			result.controller[key.Lease] = makeLeaseInfo(value)
+		case lease.ApplicationLeadershipNamespace:
+			model, ok := result.models[key.ModelUUID]
+			if !ok {
+				model = make(map[string]leaseInfo)
+				result.models[key.ModelUUID] = model
+			}
+			model[key.Lease] = makeLeaseInfo(value)
+		default:
+			logger.Warningf("unknown namespace %q", key.Namespace)
+		}
+	}
+	// TODO: handle pinned.
+	return result
+}
+
+func (h leaseHandler) filterModel(data *leases, partialModelUUID string) *leases {
+	result := &leases{
+		controller: make(map[string]leaseInfo),
+		models:     make(map[string]map[string]leaseInfo),
+	}
+	for modelUUID, value := range data.controller {
+		if strings.Contains(modelUUID, partialModelUUID) {
```

Comment:
startsWith maybe instead?
I would expect partial UUIDs would always be the first few characters. Saves from unexpected matches

Reply:
except the short hash we use almost everywhere is the last seven